### PR TITLE
Add guidelines to vocal highway

### PR DESCRIPTION
**Pull request by thomeval**
This PR adds horizontal guidelines to the background of the vocal highway, for both solo and harmony vocals. The guidelines are scaled and shifted according to the vocal range of notes on the highway. Range shifts are supported as well. In order to fix issues with notes not being aligned with the guidelines, some changes were made to the code that positions vocal notes as well.

image

Note that unlike RB3, the guidelines will always show the C-Major scale (C,E,G,C), and will ignore the song's key. This is considered out of scope for this PR, it we would need to add support for the (vocal_tonic_note) and (song_tonality) fields, which is generally only found in newer RBCON files and RBHP upgrades.

### CODE REVIEW

File: Assets/Script/Gameplay/Player/Vocals/VocalTrack.cs [hunk 1]
```diff
@@ -317,16 +335,30 @@ private void Update()
...
+        private void UpdateHighwayGuidelines()
+        {
+            const int DEFAULT_GUIDELINE_SCALE = 24;     // The semi-tone range of the guideline texture
+
+            var scale = (_viewRange.Max - _viewRange.Min) / DEFAULT_GUIDELINE_SCALE;
+            var offset = (_viewRange.Min % DEFAULT_GUIDELINE_SCALE) / DEFAULT_GUIDELINE_SCALE;
+            _soloGuidelineRenderer.material.mainTextureOffset = new Vector2(1, offset);
+            _soloGuidelineRenderer.material.mainTextureScale = new Vector2(1, scale);
+            _harmonyGuidelineRenderer.material.mainTextureOffset = new Vector2(1, offset);
+            _harmonyGuidelineRenderer.material.mainTextureScale = new Vector2(1, scale);
```
Comment (EliteAsian123) [inline Assets/Script/Gameplay/Player/Vocals/VocalTrack.cs:347]: Repeated access of the same material is quite expensive. Instead, these values should probably be cached somewhere.

File: Assets/Script/Gameplay/Player/Vocals/VocalTrack.cs [hunk 2]
```diff
@@ -41,10 +41,17 @@ public Range(float min, float max)
...
 
-        private const float TRACK_BOTTOM = -0.53f;
+        // The top edge of the vocal highway track when harmonies are enabled (excluding the lyric tracks)
+        private const float TRACK_TOP_HARMONY = 0.64f;
+
+        // The bottom edge of the vocal highway track (excluding the lyric track)
+        private const float TRACK_BOTTOM = -0.63f;
+
+        // The amount of additional padding to apply to the visible semi-tone range, expressed as a percentage.
+        private const float RANGE_PADDING_PERCENT = 0.1f;
```
Comment (EliteAsian123) [inline Assets/Script/Gameplay/Player/Vocals/VocalTrack.cs:54]: Use summary comments for these:
```
/// <summary>
/// Like this
/// </summary>
```

File: Assets/Script/Gameplay/Player/Vocals/VocalTrack.cs [hunk 3]
```diff
@@ -317,16 +335,30 @@ private void Update()
...
         {
-            // Pad out range based on note width
-            float minPitch = range.MinimumPitch - NOTE_WIDTH_MULTIPLIER / 2;
-            float maxPitch = range.MaximumPitch + NOTE_WIDTH_MULTIPLIER / 2;
 
             // Ensure range is at least a minimum size
             float rangeMiddle = (range.MaximumPitch + range.MinimumPitch) / 2;
-            float rangeMin = Math.Min(rangeMiddle - (MINIMUM_SEMITONE_RANGE / 2), minPitch);
-            float rangeMax = Math.Max(rangeMiddle + (MINIMUM_SEMITONE_RANGE / 2), maxPitch);
+            float rangeMin = Math.Min(rangeMiddle - (MINIMUM_SEMITONE_RANGE / 2), range.MinimumPitch);
```
Comment (EliteAsian123) [inline Assets/Script/Gameplay/Player/Vocals/VocalTrack.cs:355]: Remove the empty space at the top of this method

File: ProjectSettings/ProjectSettings.asset
Comment (EliteAsian123) [inline ProjectSettings/ProjectSettings.asset:1]: These file changes probably weren't intended

File: Assets/Art/Materials/Gameplay/Visual/VocalTrackGuidelines.mat
Comment (EliteAsian123) [inline Assets/Art/Materials/Gameplay/Visual/VocalTrackGuidelines.mat:1]: In my opinion, this texture is still a little too jarring. If the transparency could be toned down a bit more I think that'd be better

**Review by EliteAsian123 — CHANGES_REQUESTED** [review]
Overall pretty good!

Just a few minor nitpicks and this should be good to go!

File: Assets/Script/Gameplay/Player/Vocals/VocalTrack.cs (hunk 3 above)
Comment (thomeval) [reply Assets/Script/Gameplay/Player/Vocals/VocalTrack.cs]: This was changed due to incoming changes from the base branch, so it should be resolved now.

**Comment by thomeval** [comment]
Alright, this one is ready for another review.

**Comment by EliteAsian123** [comment]
The code looks good to me, I'll check in game for the texture tomorrow!

**Review by EliteAsian123 — APPROVED** [review]

**Pull request merged**